tests/vmi_configuration_test.go

```diff
@@ -945,8 +945,19 @@ var _ = Describe("Configurations", func() {
 				freeHugepages, err := strconv.Atoi(strings.Trim(output, "\n"))
 				Expect(err).ToNot(HaveOccurred())
 
+				output, err = tests.ExecuteCommandOnPod(
+					virtClient,
+					&pods.Items[0],
+					pods.Items[0].Spec.Containers[0].Name,
+					[]string{"cat", fmt.Sprintf("%s/resv_hugepages", hugepagesDir)},
+				)
+				Expect(err).ToNot(HaveOccurred())
+
+				resvHugepages, err := strconv.Atoi(strings.Trim(output, "\n"))
+				Expect(err).ToNot(HaveOccurred())
+
 				// Verify that the VM memory equals to a number of consumed hugepages
-				vmHugepagesConsumption := int64(totalHugepages-freeHugepages) * hugepagesSize.Value()
+				vmHugepagesConsumption := int64(totalHugepages-freeHugepages+resvHugepages) * hugepagesSize.Value()
```

Comment:
@alicefr I think we only need to care about the reserved pages. I guess we don't need this calculation.
Let's just do `vmHugepagesConsumption := int64(resvHugepages) * hugepagesSize.Value()` and it all should just work... I hope :)

Reply:
@vladikr I'm not sure. The example above https://github.com/kubevirt/kubevirt/pull/4027#issuecomment-680730415 and https://github.com/kubevirt/kubevirt/pull/4027#issuecomment-680735124 have a "wierd" page number because of the mismatch between MiB and MB. So, I was not actually using 32 MiB and 64 MiB. Let me redo the 2 examples with the right memory :):

Example with 32 Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    851968 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        6
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```

Example with 64 Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    849920 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       32
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```
Plus some extra ;) 

Example with 16Mi:
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    845824 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       56
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB
```
Example with 24 Mi
```bash
cat /proc/meminfo | grep Huge
AnonHugePages:    827392 kB
ShmemHugePages:        0 kB
HugePages_Total:      64
HugePages_Free:       54
HugePages_Rsvd:        2
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:          131072 kB

```

Reply:
Interesting :)

Actually I didn't comment based on your example.

I've printed the values from the within the test on my machine:
```
freeHugepages:  62
resvHugepages:  32
totalHugepages-freeHugepages+resvHugepages:  34
pageSize:  2097152
vmHugepagesConsumption:  71303168
vmMemory:  67108864

------------------------------
• Failure [14.329 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 71303168
        to equal
            <int64>: 67108864

```

Reply:
My assumption was that when the VMI starts, memfd will reserve the correct amount of pages for it and we test this at this point.
When I ran all the tests together with `vmHugepagesConsumption := int64(resvHugepages) * hugepagesSize.Value()` it passes, but I'm not sure if my assumption is true.

Reply:
mmh :thinking: , might be that when I tried the examples above the VM was running since a while and then the reserved pages have been allocated and when you run the tests it was still in early stage. Not sure what is the behavior.

Reply:
@vladikr did you set the memory with this change https://github.com/kubevirt/kubevirt/pull/4027#discussion_r477299771? Just to be sure

Reply:
yeah, I just pulled your latest.

Reply:
For me, that is what happens:
1-  Checking only the reserved pages when the VM is starting works because none of the pages has been allocated
2- The test that is failing is the second one because not all the hugepages from the previous test have been freed yet.
    This is why you got 34 pages and it worked only considering the reserved pages. However, as we can see from my 
    examples in the case of 64Mi all the pages have been allocated, but it is not the case for example for 32Mi.

Reply:
It will be interesting to try if we change the memory to 32Mi of the first case, and check if the second one still passes only considering the reserved pages.

Reply:
I tried to change it using only reserved pages but for me, it fails.

Reply:
oh I see... it may be only a coincidence that it worked with 64Mi
Ok, then let's follow your suggestion to wait until `free_pages==nr_pages and rserv_page = 0`
I only hope that we will get to this condition in a reasonable time so we will keep it deterministic.

Perhaps you can just wrap it in an `Eventually` statement, like in 
https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L1030

Reply:
@vladikr I'll figure out where to include the check :)

Reply:
@vladikr do you have any suggestions on how the VM creation [here](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L995) can be splitted in VM defined and then set the VM in running state later. I'd like to create the vmi pod, check if the hugepages have been freed, and then start the VM.

Reply:
alternatively, we can just try multiple times to verify the [memory consumption](https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_configuration_test.go#L1000). That is much simpler to implement. Right now, I'm having some issues with Eventually, because it doesn't retry a second time if the condition is not met. I need to verify why

Reply:
@alicefr I think verifying consumption is an easier approach. Let me try your PR locally and see what's wrong with the `Eventually`

Reply:
@vladikr I added a couple of prints locally, and the problem is that if the function `verifyHugepagesConsumption` doesn't return immediately true, then eventually doesn't retry a second time and it just times out. At least, it's what I've locally

Reply:
ah, yes, I assume it might take some time for `tests.ExecuteCommandOnPod` to execute. Perhaps, just increase the timeout (maybe even to a minute), it should help.

Reply:
Right, I've already tried to increase up to 2 min ;), but again the problem it's that it doesn't retry a second time. So, after 2 mins it times out

Reply:
I also tried to measure how much time the function takes, it's more or less 300ms.

Reply:
@vladikr without eventually, with a very basic loop it's working.
```diff
+                               for try := 0; !verifyHugepagesConsumption(); try++ {
+                                       if try == 5 {
+                                               break
+                                       }
+                                       time.Sleep(time.Second)
+                                       fmt.Printf("Try: %d \n", try)
+                               }
```

Reply:
@vladikr moving the entire function inside the `Eventually` brackets solved the issue. I hope that now it's fine.